### PR TITLE
unix,win: add uv_udp_try_send2

### PR DESCRIPTION
**Pull request by bnoordhuis**
Add a version of uv_udp_try_send that can send multiple datagrams.

Uses sendmmsg(2) on platforms that support it (Linux, FreeBSD, macOS),
falls back to a regular sendmsg(2) loop elsewhere.

This work was sponsored by ISC, the Internet Systems Consortium.

<hr>

The first commit is a NFC that is prep work for the second commit. Same line count, more functionality, improved legibility (IMO)

I moved `uv__udp_sendmsg` to the bottom of udp.c because the diff is crazy hard to read otherwise.

I don't love the name `uv_udp_try_send2` but I guess it's convention by now. Can I see hands for `uv_udp_try_sendv`? `uv_udp_try_send_many`? Other suggestions?

### CODE REVIEW

**Comment by saghul** [comment]
> don't love the name `uv_udp_try_send2` but I guess it's convention by now. Can I see hands for `uv_udp_try_sendv`? `uv_udp_try_send_many`? Other suggestions?

I like `uv_udp_try_send2` and `uv_udp_try_sendv`. The former is a more used convention, and I don't recall us having any instances of the latter, even if APIs like `uv_write` take multiple buffers, so there is that...

**Comment by bnoordhuis** [comment]
Okay, `uv_udp_try_send2` it is then. Any objections to me merging this?

**Review by saghul — APPROVED** [review]

File: docs/src/udp.rst [hunk 1]
```diff
@@ -426,6 +426,20 @@ API
 
     .. versionchanged:: 1.27.0 added support for connected sockets
 
+.. c:function:: int uv_udp_try_send2(uv_udp_t* handle, unsigned int count, uv_buf_t* bufs[/*count*/], unsigned int nbufs[/*count*/], struct sockaddr* addrs[/*count*/], unsigned int flags)
+
+    Like :c:func:`uv_udp_try_send`, but can send multiple datagrams.
+    Lightweight abstraction around :man:`sendmmsg(2)`, with a :man:`sendmsg(2)`
+    fallback loop for platforms that do not support the former. The handle must
+    be fully initialized; call c:func:`uv_udp_bind` first.
```
Comment (squeek502) [inline docs/src/udp.rst:434]: > The handle must be fully initialized; call `uv_udp_bind` first.

Could I get some clarification about this? `uv_udp_bind` is not called on the `client` handle in the updated `test-udp-try-send.c`.

**Comment by ibc** [comment]
Too late but what's the purpose of the `flags` argument?

```c
int uv_udp_try_send2(uv_udp_t* handle,
                     unsigned int count,
                     uv_buf_t* bufs[/*count*/],
                     unsigned int nbufs[/*count*/],
                     struct sockaddr* addrs[/*count*/],
                     unsigned int flags) {
  if (count < 1)
    return UV_EINVAL;

  if (flags != 0)
    return UV_EINVAL;
```

Basically if `flags != 0` then the function fails with `UV_EINVAL`. So what's the purpose of `args`?

**Comment by saghul** [comment]
Usually we add those for future usage.

**Comment by ibc** [comment]
> Usually we add those for future usage.

Makes sense. Maybe some docs stating that flags must be 0 for now would be helpful to avoid confusion (people maybe tempted to use same flags as in other uv_udp related functions).

**Comment by ibc** [comment]
Another thing I don't understand from the `uv_udp_try_send2()` function is the meaning of `count`, specially in the `addrs` array:

```c
int uv_udp_try_send2(uv_udp_t* handle,
                     unsigned int count,
                     uv_buf_t* bufs[/*count*/],
                     unsigned int nbufs[/*count*/],
                     struct sockaddr* addrs[/*count*/],
                     unsigned int flags)
```

So imagine I have a single `uv_buf_t buf` and I want to send it to `sockaddr1` and `sockaddr2` (2 different UDP destinations) from the same UDP socket in libuv. Would it be something like this (pseudo code)?

```c
uv_udp_try_send2(handle,
                 2 /*count*/,
                 [&buf, &buf] /*bufs[count]*/,
                 [1, 1] /*nbufs[count]*/,
                 [sockaddr1, sockaddr2] /*addrs[count]*/,
                 0 /*flags*/)
```

**Comment by saghul** [comment]
> So imagine I have a single `uv_buf_t buf` and I want to send it to `sockaddr1` and `sockaddr2` (2 different UDP destinations) from the same UDP socket in libuv. Would it be something like this (pseudo code)?
> 
> ```c
> uv_udp_try_send2(handle,
>                  2 /*count*/,
>                  [&buf, &buf] /*bufs[count]*/,
>                  [1, 1] /*nbufs[count]*/,
>                  [sockaddr1, sockaddr2] /*addrs[count]*/,
>                  0 /*flags*/)
> ```

Yep.

If you check the implementation: https://github.com/libuv/libuv/blob/db3d0ff33a2701971a90f9c0284d28674555f716/src/unix/udp.c#L1302 you can see that it basically sends each buffer list to the given address.

**Comment by ibc** [comment]
> If you check the implementation

Yes. The thing is: what about if I just read the docs?

**Comment by saghul** [comment]
But who does that? :-P

**Comment by ibc** [comment]
> But who does that? :-P

I tried. Then I add to dig into the PR and the test it includes.

**Pull request merged**